### PR TITLE
feat(modal): Improve modal transitions, padding adjustments, and aditional features

### PR DESCRIPTION
**Pull request by tmorehouse**
Various modal improvements:

- [x] Improve modal transitions.
- [x] Remove need for scoped CSS.
- [x] Incorporate `BvEvent` object in modal events (`hide`, `hidden`, `show`, `shown`).
- [x] Incorporate scrollbar adjustments.
- [x] Add listeners for resize/orientationchange to adjust modal padding if necessary
- [x] Use observeDom util to adjust modal when content size changes.
- [x] Option to not show backdrop (although still supports close on backdrop click, when backdrop not shown).
- [x] Emit on `$root` new namespaced events.
- [x] Listen on `$root` new namedspaced events (**Minor breaking change**).
- [x] Update `v-b-modal` directive to use new namespaced `$root` event.
- [x] Add optional bg, border, and text variants (header, footer, body).
_Note: current Bootstrap V4.beta.1 CSS causes header and footer to loose rounded corners when a background variant is applied. This PR includes a fix for that using the `rounded-*` classes_
- [x] New functional component `<b-button-close>` for header close button (also added to `<b-alert>`).
- [x] New slot for header-close button content
- [x] Change footer close button text to `Cancel` to better reflect slot name and event name
- [x] More dom utils for better minification (used by modal and tooltips/popovers).
- [x] Auto append tooltip/popovers to `.modal` `<div>` (instead of `<body>`) for proper z-indexing, if container not specified on tooltip/popover (check is done within tooltip.js class)
- [x] Update docs & examples.

### CODE REVIEW

**Comment by tmorehouse** [comment]
I think everything is now ready for the revamped modal. component

**Comment by tmorehouse** [comment]
Added additional dom utils methods for better minification

**Comment by tmorehouse** [comment]
OK... now I think this PR is ready for use.

**Comment by jackmu95** [comment]
@tmorehouse Nice work!

**Comment by alexsasharegan** [comment]
It would be nice to work all the examples (turn most into live play ex), but I think this is good now. Examples could be another pr.

File: lib/components/modal.vue [hunk 1]
```diff
@@ -249,49 +249,61 @@
...
                     // Rounding is needed to fix a bug in bootstrap V4.beta.1 CSS
-                    Boolean(this.headerBgVariant) ? 'rounded-top' : '',
-                    Boolean(this.headerBgVariant) ? `bg-${this.headerBgVariant}` : '',
-                    Boolean(this.headerTextVariant) ? `text-${this.headerTextVariant}` : '',
-                    Boolean(this.headerBorderVariant) ? `border-${this.headerBorderVariant}` : ''
+                    {
+                        'rounded-top': this.headerBgVariant,
+                        [`bg-${this.headerBgVariant}`]: this.headerBgVariant,
+                        [`text-${this.headerTextVariant}`]: this.headerTextVariant,
+                        [`border-${this.headerBorderVariant}`]: this.headerBorderVariant,
```
Comment (tmorehouse) [inline lib/components/modal.vue]: Should these be wrapping the variants in `Boolean(...)` (since they are strings)

Comment (tmorehouse) [reply lib/components/modal.vue]: i.e
```js
[`bg-${this.headerBgVariant}`]: Boolean(this.headerBgVariant),
```

Comment (tmorehouse) [reply lib/components/modal.vue]: And similar for the other classes sections as well...

Comment (alexsasharegan) [reply lib/components/modal.vue]: Like we discussed, Vue checks these in an `if` condition, so it isn't necessary. You're thought and them getting cached in the computed props is smart though, so best to cast them.

**Pull request merged**